src/lib/statistics/statistics_objects/top_k_uniform_distribution_histogram.cpp

```diff
@@ -0,0 +1,154 @@
+  Assert(max_bin_count > 0, "max_bin_count must be greater than zero ");
+
+  auto value_distribution = value_distribution_from_column(table, column_id, domain);
+
+  if (value_distribution.empty()) {
+    return nullptr;
+  }
+
+  // If the column holds less than K distinct values use the distinct count as TOP_K instead
+
+  const auto k = std::min(TOP_K_DEFAULT, value_distribution.size());
+
+  // Get the first top k values and save them into vectors
+  std::vector<T> top_k_names(k);
+  std::vector<HistogramCountType> top_k_counts(k);
+
+  // Sort values by occurrence count
+  auto sorted_count_values = value_distribution;
+  std::sort(sorted_count_values.begin(), sorted_count_values.end(),
+            [&](const auto& l, const auto& r) { return l.second > r.second; });
+
+  // Sort TOP_K values with highest occurrence count lexicographically.
+  // We later use this for more performant range predicate evaluation.
+  std::sort(sorted_count_values.begin(), sorted_count_values.begin() + k,
+            [&](const auto& l, const auto& r) { return l.first < r.first; });
+
+  if (!sorted_count_values.empty()) {
+    for(auto i = 0u; i < k; i++) {
+      top_k_names[i] = sorted_count_values[i].first;
+      top_k_counts[i] = sorted_count_values[i].second;
+    }
+  }
+
+  // Remove TOP_K values from value distribution
+  for (auto i = 0u; i < k; i++) {
+    auto it = remove(value_distribution.begin(), value_distribution.end(), std::make_pair(top_k_names[i], top_k_counts[i]));
+    value_distribution.erase(it, value_distribution.end());
+  }
+
+  // Model uniform distribution as one bin for all non-Top K values.
+  const auto bin_count = value_distribution.size() < 1 ? BinID{top_k_names.size()} : BinID{2*top_k_names.size() + 1};
+
+  GenericHistogramBuilder<T> builder{bin_count, domain};
+
+  // Split values evenly among bins.
+  //const auto range_minimum = value_distribution.front().first;
+  
+  const auto non_top_k_count = std::accumulate(
+        value_distribution.cbegin(), 
+        value_distribution.cend(), 
+        HistogramCountType{0}, 
+        [](HistogramCountType a, const std::pair<T, HistogramCountType>& b) { return a + b.second; });
+
+  const auto bin_distinct_count = value_distribution.size();
+  const auto count_per_non_top_k_value = bin_distinct_count != 0 ? non_top_k_count / bin_distinct_count : BinID{0};
+
+  auto current_minimum_index = 0u;
+  auto current_maximum_index = value_distribution.size() - 1;
+
+  for (auto top_k_index = 0u, top_k_size = top_k_names.size(); top_k_index < top_k_size; top_k_index++) {
+    // current top_k value
+    auto skip_bin = false;
+    const auto current_top_k_value = top_k_names[top_k_index];
+
+    // find value smaller than current top_k value for current maximum
+    auto value_dist_lower_bound = std::lower_bound(value_distribution.begin(), value_distribution.end(), current_top_k_value,
```

Comment:
Do we need to search from `begin()` to `end()`? Is it possible to search a sub-part of the array?